README.pod
lib/WebDriver/Tiny.pod

Changes
{{ $NEXT }}

0.105     2021-04-15 12:15:03+01:00 Europe/London
    - Change author in META. No other code changes

0.104     2021-03-31 12:20:35+01:00 Europe/London
    - Bump version in WebDriver::Tiny::Elements, which lagged behind.

0.103     2020-06-21 18:49:25+01:00 Europe/London
    - Remove documentation for methods removed previously.

0.102 2019-05-17
    - Continue to support the latest WebDriver spec as it evolves, older
      servers may no longer work.

0.101 2018-01-02
    - Try harder to get a meaningful error message.
    - Make use of subroutine signatures.

0.100 2017-12-21
    - This is a breaking change, we now support the latest WebDriver spec and
      test on the latest geckodriver, I suspect phantomjs no longer works, but
      chromedirver should.
    - Add $elem->prop.
    - Drop $drv->js_phantom.
    - Drop $drv->storage.
    - Drop $elem->location.
    - Drop $elem->location_in_view.
    - Drop $elem->move_to.
    - Drop $elem->submit.
    - Rename $drv->*_alert to $drv->alert_*.
    - Rename $drv->source to $drv->html, add $elem->html.
    - Replace $drv->window_{postion,size} with $drv->window_rect.

0.006 2016-12-14
    - Reorganise window methods.
    - Always send a JSON payload, to appease geckodriver.
    - Assorted fixes. Thanks arc.

0.005 2016-11-22
    - Doc fixes. Thanks tomhukins.
    - Avoid timeouts by using HTTP::Tiny's public API.

0.004 2016-02-24
    - Require Perl 5.20 to allow for cleaner code.
    - Require Carp 1.25 to ensure errors have periods on the end.
    - Add $drv->alert_text.
    - Add $drv->status.
    - Add $drv->storage.
    - Add $elem->location.
    - Add $elem->location_in_view.

0.003 2015-09-11
    - Add support for ChromeDriver.
    - Document how to use different backends.
    - Add cookie getter/setter.
    - Rename delete_cookie to cookie_delete.
    - Support cookie_delete(@list_of_cookies_to_delete).
    - Add support for asynchronous JavaScript.
    - Rename execute to js.
    - Rename execute_phantom to js_phantom.
    - Add append, split, and uniq methods to element collections.

0.002 2015-07-05
    - $drv->() caches the closure and is now twice as fast.
    - $drv->base_url('foo'), and $drv->window_maximize now return $self.

0.001 2015-06-09
    - Initial release.


t/t.pm
package t;

use strict;
use warnings;

require WebDriver::Tiny;

sub import {
    # Turn on strict & warnings for the caller.
    strict->import;
    warnings->import;

    eval "package main; use Test::More tests => $_[1]";
}

my @reqs;

# Give the caller a reqs_are test sub.
sub main::reqs_are {
    Test::More::is_deeply( \@reqs, @_ );

    @reqs = ();
}

# Give the caller a $content so they can override.
*main::content = \( my $content = '{}' );

# Our dummy user agent just logs what the request was and returns success.
sub ua::request {
    shift;

    # Decode JSON, if provided, to make testing easier.
    $_[2] = JSON::PP::decode_json( $_[2]{content} ) if $_[2];

    push @reqs, [@_];

    { content => $content, success => 1 }
};

my $drv = bless [ bless( [], 'ua' ), '', '' ], 'WebDriver::Tiny';

$drv->[4] = sub { $drv->find(@_) };

# Give the caller a $drv and a $elem.
*main::drv = \$drv;

*main::elem = \bless [ $drv, 123 ], 'WebDriver::Tiny::Elements';


xt/form.pl
use strict;
use utf8;
use warnings;

use URI;
use URI::QueryParam;

sub {
    my $drv = shift;

    is_deeply [ map $_->attr('name'), $drv->('input,select') ], [
        'text', "text '", 'text "', 'text \\', 'text ☃',
        ('radio') x 3, 'select', 'multi select',
    ], 'names of all form fields are correct';

    # Perl 6 envy :-(
    sub pick { map { splice @_, rand @_, 1 } 1 .. shift }
    sub roll { map { $_[ rand @_ ]         } 1 .. shift }

    $drv->('form')->submit( my %values = (
        'text'         => join( '', roll 3, 'a'..'z' ),
        "text '"       => join( '', roll 3, 'a'..'z' ),
        'text "'       => join( '', roll 3, 'a'..'z' ),
        'text \\'      => join( '', roll 3, 'a'..'z' ),
        'text ☃'      => join( '', roll 3, 'a'..'z' ),
        'radio'        => roll( 1, 'a'..'c' ),
        'select'       => roll( 1, 'a'..'c' ),
        'multi select' => [ pick( 2, 'a'..'c' ) ],
    ) );

    my %expected;

    while ( my ( $k, $v ) = each %values ) {
        utf8::encode $k;

        $expected{$k} = ref $v ? bag @$v : $v;
    }

    cmp_deeply +URI->new( $drv->url )->query_form_hash, \%expected,
        'submit works on all form fields correctly';

    my $elem = $drv->('input[type=text]');

    $elem->send_keys('♥ ♦ ♣ ♠');

    is $elem->attr('value'), '♥ ♦ ♣ ♠', 'elem->send_keys';

    $elem->clear;

    is $elem->attr('value'), '', 'elem->clear';
};


xt/basic.pl
use strict;
use utf8;
use warnings;

sub {
    my $drv = shift;

    # like $drv->window, qr/^[-\w]+$/, 'window';

    cmp_deeply $drv->windows, [ re qr/^[-\w]+$/ ], 'windows';

    my $got = $drv->html;

    # Try to normalise the HTML.
    $got =~ s/\s//g;
    $got =~ s/\/>/>/g;
    $got =~ s|xmlns="http://www.w3.org/1999/xhtml"||;

    is $got, $::html =~ s/\s//gr, 'source';

    is $drv->('title')->html, '<title>Frosty the ☃</title>', 'html of <title>';

    is $drv->('h1')->text, 'ᴛ̲ʜ̲ᴇ̲ʀ̲ᴇ̲ ̲ɪ̲s̲ ̲ɴ̲ᴏ̲ ̲U̲ɴ̲ɪ̲ᴄ̲ᴏ̲ᴅ̲ᴇ̲ ̲ᴍ̲ᴀ̲ɢ̲ɪ̲ᴄ̲ ̲ʙ̲ᴜ̲ʟ̲ʟ̲ᴇ̲ᴛ̲', 'text';

    is $drv->('h3')->text, 'foo bar', 'text on more than one element';

    is_deeply [ map $_->text, $drv->('h3') ], [qw/foo bar/],
        'find in list context';

    is_deeply [ map $_->text, $drv->('h3')->split ], [qw/foo bar/],
        'split';

    is $drv->title, 'Frosty the ☃', 'title';

    is $drv->url, 'http://localhost:8080/', 'url';

    ( my $bottom = $drv->( 'go to bottom', method => 'link_text' ) )->click;

    is $drv->url, 'http://localhost:8080/#bottom', 'click';

    ( my $top = $drv->( 'go to top', method => 'link_text' ) )->click;

    is $drv->url, 'http://localhost:8080/#top', 'click';

    is_deeply $drv->( 'to top', method => 'partial_link_text' ), $top,
        'partial_link_text matching one';

    is_deeply [ $drv->( ' to ', method => 'partial_link_text' ) ],
        [ $bottom, $top ], 'partial_link_text matching two';
}


lib/WebDriver/Tiny/Elements.pm
package WebDriver::Tiny::Elements 0.104;

use 5.020;
use feature qw/postderef signatures/;
use warnings;
no  warnings 'experimental';

# Manip
sub append { bless [ shift->@*, map @$_[ 1.. $#$_ ], @_ ] }
sub first  { bless [ $_[0]->@[ 0,  1 ] ] }
sub last   { bless [ $_[0]->@[ 0, -1 ] ] }
sub size   { $#{ $_[0] } }
sub slice  { my ( $drv, @ids ) = shift->@*; bless [ $drv, @ids[@_] ] }
sub split  { my ( $drv, @ids ) = $_[0]->@*; map { bless [ $drv, $_ ] } @ids }

sub uniq($self) {
    my ( $drv, @ids ) = @$self;

    bless [ $drv, keys %{ { map { $_ => undef } @ids } } ];
}

sub attr($self, $value) { $self->_req( GET => "/attribute/$value" ) }
sub  css($self, $value) { $self->_req( GET =>       "/css/$value" ) }
sub prop($self, $value) { $self->_req( GET =>  "/property/$value" ) }

sub clear($self) { $self->_req( POST => '/clear' ); $self }
sub click($self) { $self->_req( POST => '/click' ); $self }
sub   tap($self) { $self->_req( POST => '/tap'   ); $self }

sub  enabled($self) { $self->_req( GET => '/enabled'   ) }
sub     rect($self) { $self->_req( GET => '/rect'      ) }
sub selected($self) { $self->_req( GET => '/selected'  ) }
sub      tag($self) { $self->_req( GET => '/name'      ) }
sub  visible($self) { $self->_req( GET => '/displayed' ) }

sub html { $_[0][0]->js( 'return arguments[0].outerHTML', $_[0] ) }

*find = \&WebDriver::Tiny::find;

sub screenshot($self, $file = undef) {
    require MIME::Base64;

    my $data = MIME::Base64::decode_base64(
        $self->_req( GET => '/screenshot' )
    );

    if ( defined $file ) {
        open my $fh, '>', $file or die $!;
        print $fh $data;
        close $fh or die $!;

        return $self;
    }

    $data;
}

sub send_keys($self, $keys) {
    $self->_req( POST => '/value', { text => "$keys" } );
    $self;
}

sub text($self) {
    my ( $drv, @ids ) = @$self;

    join ' ', map $drv->_req( GET => "/element/$_/text" ), @ids;
}

# Call driver's ->_req, prepend "/element/:id" to the path first.
sub _req { $_[0][0]->_req( $_[1], "/element/$_[0][1]$_[2]", $_[3] ) }

1;


lib/WebDriver/Tiny.pm
package WebDriver::Tiny 0.104;

use 5.020;
use feature qw/lexical_subs postderef signatures/;
use warnings;
no  warnings 'experimental';

# https://www.w3.org/TR/webdriver/#elements
my sub ELEMENT_ID :prototype() {'element-6066-11e4-a52e-4f735466cecf'}

# Allow "cute" $drv->('selector') syntax.
#
# $self->[4] contains a sub ref that closes over a weakened $self that calls
# find, i.e. sub { $weak_self->find(@_) }. This sub ref is returned when $self
# is invoked as a sub ref thanks to the magic of overloading. We weaken to
# avoid a memory leak. The closure is built in new().
use overload fallback => 1, '&{}' => sub { $_[0][4] };

use Carp 1.25 ();
use HTTP::Tiny;
use JSON::PP ();
use WebDriver::Tiny::Elements;

our @CARP_NOT = 'WebDriver::Tiny::Elements';

sub import {
    # https://www.w3.org/TR/webdriver/#sendkeys
    state $chars = {
        WD_NULL            => 57344, WD_CANCEL     => 57345,
        WD_HELP            => 57346, WD_BACK_SPACE => 57347,
        WD_TAB             => 57348, WD_CLEAR      => 57349,
        WD_RETURN          => 57350, WD_ENTER      => 57351,
        WD_SHIFT           => 57352, WD_CONTROL    => 57353,
        WD_ALT             => 57354, WD_PAUSE      => 57355,
        WD_ESCAPE          => 57356, WD_SPACE      => 57357,
        WD_PAGE_UP         => 57358, WD_PAGE_DOWN  => 57359,
        WD_END             => 57360, WD_HOME       => 57361,
        WD_ARROW_LEFT      => 57362, WD_ARROW_UP   => 57363,
        WD_ARROW_RIGHT     => 57364, WD_ARROW_DOWN => 57365,
        WD_INSERT          => 57366, WD_DELETE     => 57367,
        WD_SEMICOLON       => 57368, WD_EQUALS     => 57369,
        WD_NUMPAD0         => 57370, WD_NUMPAD1    => 57371,
        WD_NUMPAD2         => 57372, WD_NUMPAD3    => 57373,
        WD_NUMPAD4         => 57374, WD_NUMPAD5    => 57375,
        WD_NUMPAD6         => 57376, WD_NUMPAD7    => 57377,
        WD_NUMPAD8         => 57378, WD_NUMPAD9    => 57379,
        WD_MULTIPLY        => 57380, WD_ADD        => 57381,
        WD_SEPARATOR       => 57382, WD_SUBTRACT   => 57383,
        WD_DECIMAL         => 57384, WD_DIVIDE     => 57385,
        WD_F1              => 57393, WD_F2         => 57394,
        WD_F3              => 57395, WD_F4         => 57396,
        WD_F5              => 57397, WD_F6         => 57398,
        WD_F7              => 57399, WD_F8         => 57400,
        WD_F9              => 57401, WD_F10        => 57402,
        WD_F11             => 57403, WD_F12        => 57404,
        WD_META            => 57405, WD_COMMAND    => 57405,
        WD_ZENKAKU_HANKAKU => 57408,
    };

    require charnames;

    charnames->import( ':alias' => $chars );
}

# We're a blessed arrayref (for speed) of the following parts:
#
# $self = [
#   0: HTTP::Tiny instance,
#   1: URL of the WebDriver daemon,
#   2: Base URL which schemeless get calls are based off,
#   3: The capabilities of the WebDriver daemon,
#   4: Cached closure of $self for ->() syntax,
# ]
sub new($class, %args) {
    Carp::croak qq/$class - Missing required parameter "port"/
        unless exists $args{port};

    $args{host} //= 'localhost';
    $args{path} //= '';

    my $self = bless [
        HTTP::Tiny->new,
        "http://$args{host}:$args{port}$args{path}/session",
        $args{base_url} // '',
    ], $class;

    my $reply = $self->_req(
        POST => '',
        { capabilities => { alwaysMatch => $args{capabilities} // {} } },
    );

    $self->[1] .= '/' . $reply->{sessionId};

    # Store the capabilities.
    $self->[3] = $reply->{capabilities};

    # Numify bool objects, saves memory.
    $_ += 0 for grep ref eq 'JSON::PP::Boolean', values $self->[3]->%*;

    # See the overloading at the top of the file for details.
    Scalar::Util::weaken( my $weak_self = $self );
    $self->[4] = sub { $weak_self->find(@_) };

    $self;
}

sub capabilities($self) { $self->[3] }

sub  html($self) { $self->_req( GET => '/source' ) }
sub title($self) { $self->_req( GET => '/title'  ) }
sub   url($self) { $self->_req( GET => '/url'    ) }

sub    back($self) { $self->_req( POST => '/back'    ); $self }
sub forward($self) { $self->_req( POST => '/forward' ); $self }
sub refresh($self) { $self->_req( POST => '/refresh' ); $self }

sub status {
    # /status is the only path without the session prefix, so surpress it.
    local $_[0][1] = substr $_[0][1], 0, rindex $_[0][1], '/session/';

    $_[0]->_req( GET => '/status' );
}

sub  alert_accept($self) { $self->_req( POST => '/alert/accept'  ); $self }
sub alert_dismiss($self) { $self->_req( POST => '/alert/dismiss' ); $self }

sub alert_text($self) { $self->_req( GET => '/alert/text' ) }

sub base_url {
    if ( @_ == 2 ) {
        $_[0][2] = $_[1] // '';

        return $_[0];
    }

    $_[0][2];
}

sub cookie {
    my ( $self, $name, $value, @args ) = @_;

    # GET /cookie/{name} isn't supported by ChromeDriver, so get all.
    return $self->cookies->{$name} if @_ == 2;

    $self->_req( POST => '/cookie',
        { cookie => { name => $name, value => $value, @args } } );

    $self;
}

sub cookie_delete($self, @cookies) {
    if (@cookies) {
        $self->_req( DELETE => "/cookie/$_" ) for @cookies;
    }
    else {
        $self->_req( DELETE => '/cookie' );
    }

    $self;
}

sub cookies {
    my @cookies = @{ $_[0]->_req( GET => '/cookie' ) // [] };

    # Map the incorrect key to the correct key.
    $_->{httpOnly} //= delete $_->{httponly} for @cookies;

    +{ map { $_->{name} => $_ } @cookies };
}

# NOTE This method can be called from a driver or a collection of elements.
sub find($self, $selector, %args) {
    state $methods = {
        css               => 'css selector',
        ecmascript        => 'ecmascript',
        link_text         => 'link text',
        partial_link_text => 'partial link text',
        xpath             => 'xpath',
    };

    my $method = $methods->{ $args{method} // '' } // 'css selector';

    # FIXME
    my $drv = ref $self eq 'WebDriver::Tiny::Elements' ? $self->[0] : $self;

    my @ids;

    for ( 0 .. ( $args{tries} // 5 ) ) {
        my $reply = $self->_req(
            POST => '/elements',
            { using => $method, value => "$selector" },
        );

        @ids = map $_->{+ELEMENT_ID}, @$reply;

        @ids = grep {
            $drv->_req( GET => "/element/$_/displayed" )
        } @ids if $args{visible};

        last if @ids;

        select undef, undef, undef, $args{sleep} // .1;
    }

    Carp::croak ref $self, qq/->find failed for $method = "$_[1]"/
        if !@ids && !exists $args{dies} && !$args{dies};

    wantarray ? map { bless [ $drv, $_ ], 'WebDriver::Tiny::Elements' } @ids
              : bless [ $drv, @ids ], 'WebDriver::Tiny::Elements';
}

my $js = sub($path, $self, $script, @args) {
    # Currently only takes the first ID in the collection, this should change.
    $_ = { ELEMENT_ID, $_->[1] }
        for grep ref eq 'WebDriver::Tiny::Elements', @args;
    $self->_req( POST => $path, { script => $script, args => \@args } );
};

sub js       { unshift @_, '/execute/sync';  goto $js }
sub js_async { unshift @_, '/execute/async'; goto $js }

sub get($self, $url) {
    $self->_req(
        POST => '/url',
        { url => $url =~ m(^https?://) ? $url : $self->[2] . $url },
    );

    $self;
}

my $mime_data = sub ($method, $path, $self, $file, %args) {
    require MIME::Base64;

    my $data = MIME::Base64::decode_base64(
        $self->_req( $method, $path, %args ? \%args : () )
    );

    if ( defined $file ) {
        open my $fh, '>', $file or die $!;
        print $fh $data;
        close $fh or die $!;

        return $self;
    }

    $data;
};

sub print {
    my $self = shift;
    unshift @_, undef unless @_ % 2;
    unshift @_, 'POST' => '/print', $self;
    goto $mime_data;
}

sub screenshot {
    push @_, undef if @_ == 1;
    unshift @_, 'GET' => '/screenshot';
    goto $mime_data;
}

sub user_agent($self) { $js->( '/execute/sync', $self, 'return window.navigator.userAgent') }

sub  window($self) { $self->_req( GET => '/window'         ) }
sub windows($self) { $self->_req( GET => '/window/handles' ) }

sub      window_close($self) { $self->_req( DELETE => '/window'            ); $self }
sub window_fullscreen($self) { $self->_req( POST   => '/window/fullscreen' ); $self }
sub   window_maximize($self) { $self->_req( POST   => '/window/maximize'   ); $self }
sub   window_minimize($self) { $self->_req( POST   => '/window/minimize'   ); $self }

sub window_rect {
    my $self = shift;

    return $self->_req( GET => '/window/rect' ) unless @_;

    $#_ = 3;

    my %args;
    @args{ qw/width height x y/ } = map $_ // 0, @_;

    $self->_req( POST => '/window/rect', \%args );

    $self;
}

sub window_switch($self, $handle) {
    $self->_req( POST => '/window', { handle => $handle } );

    $self;
}

sub _req($self, $method, $path, $args = undef) {
    my $reply = $self->[0]->request(
        $method,
        $self->[1] . $path,
        { content => JSON::PP::encode_json( $args // {} ) },
    );

    my $value = eval { JSON::PP::decode_json( $reply->{content} )->{value} };

    unless ( $reply->{success} ) {
        my $error = $value
            ? $value->{message} || $value->{error} || $reply->{content}
            : $reply->{content};

        Carp::croak ref $self, ' - ', $error;
    }

    $value;
}

sub DESTROY($self) { $self->_req( DELETE => '' ) if $self->[0] && $self->[3] }

1;
